### PR TITLE
SONARJAVA-1918

### PR DESCRIPTION
**Pull request by ganncamp**
Initial implementation

### CODE REVIEW

File: java-checks/src/main/java/org/sonar/java/checks/CognitiveComplexityMethodCheck.java [hunk 1]
```diff
@@ -0,0 +1,476 @@
...
+        break;
+      default:
+        break;
+    }
+
+    return st;
+  }
+
+
+  private List<StatementTree> getChildren(Tree st) {
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/CognitiveComplexityMethodCheck.java]: CRITICAL The Cyclomatic Complexity of this method "getChildren" is 20 which is greater than 10 authorized. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
MAJOR Make "getChildren" a "static" method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

File: java-checks/src/main/java/org/sonar/java/checks/CognitiveComplexityMethodCheck.java [hunk 2]
```diff
@@ -0,0 +1,476 @@
...
+
+      if (tryStatment.finallyBlock() != null && tryStatment.finallyBlock().body() != null) {
+        total += countComplexity(tryStatment.finallyBlock().body(), flow, nestingLevel);
+      }
+    }
+
+    return total;
+  }
+
+  private int countBreakAndContinue(Tree st, List<JavaFileScannerContext.Location> flow) {
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/CognitiveComplexityMethodCheck.java]: MAJOR Make "countBreakAndContinue" a "static" method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

File: java-checks/src/main/java/org/sonar/java/checks/CognitiveComplexityMethodCheck.java [hunk 3]
```diff
@@ -0,0 +1,476 @@
...
+   * Increment for each non-like operator. So:
+   *   a           // +0
+   *   a && b      // +1
+   *   a && b && c // +1
+   *   a && b || c // +2
+   *
+   * @param expressionTree
+   * @return
+   */
+  private int countConditions(ExpressionTree expressionTree, List<JavaFileScannerContext.Location> flow) {
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/CognitiveComplexityMethodCheck.java]: CRITICAL The Cyclomatic Complexity of this method "countConditions" is 13 which is greater than 10 authorized. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: java-checks/src/main/java/org/sonar/java/checks/CognitiveComplexityMethodCheck.java [hunk 4]
```diff
@@ -0,0 +1,476 @@
...
+    switch (st.kind()) {
+      case LABELED_STATEMENT:
+        return ((LabeledStatementTree)st).statement();
+      case VARIABLE:
+        VariableTree vt = (VariableTree)st;
+        if (vt.initializer() != null) {
+          return vt.initializer();
+        }
+        break;
+      case EXPRESSION_STATEMENT:
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/CognitiveComplexityMethodCheck.java]: MAJOR Reduce this switch case number of lines from 7 to at most 5, for example by extracting code into methods. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1151)

**Comment by SonarTech** [comment]
SonarQube analysis reported 10 issues
* CRITICAL 4 critical
* MAJOR 6 major

Watch the comments in this conversation to review them.

File: java-checks/src/main/java/org/sonar/java/checks/CognitiveComplexityMethodCheck.java [hunk 5]
```diff
@@ -0,0 +1,476 @@
...
+        total += countComplexity(getChildren(tree), flow, nestingLevel);
+      } else {
+        total += countComplexity(getChildren(tree), flow, nestingLevel + 1);
+      }
+    }
+
+    return total;
+  }
+
+  private static void addSecondaryLocation(List<JavaFileScannerContext.Location> flow, Tree st, int hit,  int nestingLevel) {
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/CognitiveComplexityMethodCheck.java]: CRITICAL The Cyclomatic Complexity of this method "addSecondaryLocation" is 14 which is greater than 10 authorized. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: java-checks/src/main/java/org/sonar/java/checks/CognitiveComplexityMethodCheck.java [hunk 6]
```diff
@@ -0,0 +1,476 @@
...
+            || (st.is(CONTINUE_STATEMENT) && ((ContinueStatementTree) st).label() != null) ) {
+      flow.add(new JavaFileScannerContext.Location("+1", st));
+      return 1;
+
+    }
+
+    return 0;
+  }
+
+  private static Tree getSegmentOfInterest(StatementTree st) {
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/CognitiveComplexityMethodCheck.java]: CRITICAL The Cyclomatic Complexity of this method "getSegmentOfInterest" is 16 which is greater than 10 authorized. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues
* CRITICAL 4 critical
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: java-checks/src/main/java/org/sonar/java/checks/CognitiveComplexityMethodCheck.java [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+  public void visitNode(Tree tree) {
+    MethodTree method = (MethodTree) tree;
+    if (method.block() == null || ((ClassTree)method.parent()).simpleName() == null || isWithinLocalClass(method)) {
+      return;
+    }
+    CognitiveComplexityVisitor complexityVisitor = new CognitiveComplexityVisitor();
+    method.accept(complexityVisitor);
+    int total = complexityVisitor.complexity;
+    if (total > max) {
+      reportIssue(method.simpleName(), "Refactor this method to reduce its Cognitive Complexity from " + total + " to the " + max + " allowed.", complexityVisitor.flow, total - max);
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/CognitiveComplexityMethodCheck.java]: MINOR Split this 182 characters long line (which is greater than 180 authorized). [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS00103)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* MINOR 1 minor

Watch the comments in this conversation to review them.

File: java-checks/src/main/resources/org/sonar/l10n/java/rules/squid/S3776_java.json [hunk 8]
```diff
@@ -0,0 +1,15 @@
+{
+  "title": "Cognitive Complexity of methods should not be too high",
```
Comment (ganncamp) [inline java-checks/src/main/resources/org/sonar/l10n/java/rules/squid/S3776_java.json]: I generated this before putting the rule in the default profile in RSpec. Will there be a final re-generation before release @benzonico?

File: java-checks/src/main/java/org/sonar/java/checks/CognitiveComplexityMethodCheck.java [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+    int nesting;
+
+    CognitiveComplexityVisitor() {
+      complexity = 0;
+      nesting = 1;
+      flow = new ArrayList<>();
+      ignored = new HashSet<>();
+    }
+
+    private void increaseComplexity(Tree tree, int increase) {
```
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/CognitiveComplexityMethodCheck.java]: This method is currently always used to increase the complexity by the `nesting` value, or by `1`. It probably worth to have two distinguished methods to cover these 2 cases explicitly.

File: java-checks/src/main/java/org/sonar/java/checks/CognitiveComplexityMethodCheck.java [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+    @Override
+    public void visitLambdaExpression(LambdaExpressionTree lambdaExpressionTree) {
+      nesting++;
+      super.visitLambdaExpression(lambdaExpressionTree);
+      nesting--;
+    }
+
+    @Override
+    public void visitBinaryExpression(BinaryExpressionTree tree) {
+      if(tree.is(CONDITIONAL_AND, CONDITIONAL_OR)) {
```
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/CognitiveComplexityMethodCheck.java]: I would also consider the `XOR` (`^`) operator when the expression is a boolean expression.

File: java-checks/src/main/resources/org/sonar/l10n/java/rules/squid/S3776_java.json (hunk 8 above)
Comment (benzonico) [inline java-checks/src/main/resources/org/sonar/l10n/java/rules/squid/S3776_java.json]: I'll update the description before merging.

File: java-checks/src/main/java/org/sonar/java/checks/CognitiveComplexityMethodCheck.java (hunk 9 above)
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/CognitiveComplexityMethodCheck.java]: fixed.

File: java-checks/src/main/java/org/sonar/java/checks/CognitiveComplexityMethodCheck.java [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.TryStatementTree;
+import org.sonar.plugins.java.api.tree.WhileStatementTree;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import static org.sonar.plugins.java.api.tree.Tree.Kind.CONDITIONAL_AND;
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/CognitiveComplexityMethodCheck.java]: MAJOR Reduce the number of "static" imports in this class from 5 to the maximum allowed 4. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS3030)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* MAJOR 1 major
* MINOR 1 minor

Watch the comments in this conversation to review them.

**Comment by ganncamp** [comment]
@benzonico I just realized that I completely forgot to implement the technical debt for this rule. Did you remember?

**Pull request merged**